### ISSUE TITLE
Investigate the perf drop on timm for dynamic shape when layout optimization is enabled

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#102670]
### 🚀 The feature, motivation and pitch

We see 14% perf win on timm models with layout optimization. But enabling dynamic shape cause 6% perf drop instead.

Theoretically we should also see wins with dynamic shape.

I've done some early investigation on gernet_l.

Here are the perf numbers:
static shape + layout opt: 1.474x
static shape without layout opt: 1.165x

dyn shape + layout opt: 0.912x
dyn shape without layout opt: 1.102x

Dive into the dynamic shape runs further.
For dyn shape + layout opt
- Abs latency: 80.241734 ms
- Fwd wrapper: https://gist.github.com/shunting314/147285c6ba55576ad528b6bc3c02aff7
- Fwd wrapper profiling result: ​​https://gist.github.com/shunting314/dcb6d8a80739487a8a782f84a3c01976 
- Fwd wrapper total wall time 28.636 ms
- Bwd wrapper: https://gist.github.com/shunting314/6218248272f9519d497a91403b676103 
- Bwd wrapper profiling result: https://gist.github.com/shunting314/d7afef45fcfa7e5b7b776b61ed9fa7a2 
- Bwd wrapper Total wall time 49.491 ms

For dyn shape without layout opt:
- Abs latency: 66.486952 ms
- Fwd wrapper: https://gist.github.com/shunting314/f0a0c038850a94283ae3a87975e75e2b 
- Fwd wrapper profiling result: https://gist.github.com/shunting314/e649dc72df2c9a247952d91e60a8de10
- Fwd wrapper Total wall time 20.838 ms
- Bwd wrapper: https://gist.github.com/shunting314/5dbc4ae04c1b8c695d31651af6185509 
- Bwd wrapper profiling result: https://gist.github.com/shunting314/41adfa37fd1cde3600f40ffebe4532fc
- Bwd wrapper Total wall time 44.032 ms 

We can see that with layout opt, overall we slow down by ~14ms. While the fwd wrapper slows downs by ~8ms and bwd wrapper slows down by ~5.5 ms. I'll dive into the fwd wrapper as the next step.

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel @jansel @Chillee @eellison

**Comment by shunting314** [comment]
By checking the profile for forward graph, we can see that enabling layout optimization still reduce cutlass/cublas/cudnn kernels (check 'unknown category kernels' section in the profile) from 11.4ms to 6.89ms. However, we increases reduction kernel time from 5.85ms to 16.85ms!

From offline discussion with @jansel, he mentioned that this might be related to split reduction. And here are experiments that confirm the hypothesis:

I tried to run experiments for static shape with or w/o split reductions.

Static shape without layout optimization: 1.165x
Static shape with layout optimization: 1.474x  (this shows speed up from layout optimization)

Static shape without layout optimization, without split reduction: 1.153x
Static shape with layout optimization, without split reduction: 0.899x

This shows that with layout optimization, we get much more penalty from disabling split reduction. Reduction in batch norm may run much faster with contiguous inputs compared to channels last when split reduction is disabled.

With dynamic shape, we disable split reduction implicitly. Thus we see similar behavior as manually disabling split reduction for static shape case.

**Comment by shunting314** [comment]
@ezyang , @voznesenskym  did you tried enabling split reduction for dynamic shape before? Could you share your findings?

**Comment by ngimel** [comment]
We should get smaller perf penalty on split reductions once #102486 lands, that would result in a single kernel for bn as opposed to 2 (with epilogues) now. 
Also, yes, enabling split reduction will be beneficial, but we'd need to somehow make assumptions on the splits for dynamic sizes.

**Comment by masnesral** [comment]
[old issue scrubbing] I believe this is no longer a problem. Here's benchmark run for timm: https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Sat%2C%2005%20Jul%202025%2015%3A32%3A06%20GMT&stopTime=Sat%2C%2012%20Jul%202025%2015%3A32%3A06%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(h100)&lBranch=gh/masnesral/219/head&lCommit=1f76fb8146b71a803a160a63f84ff4b0d49c9ed9&rBranch=gh/masnesral/220/head&rCommit=9f4ec96ad37cf0683ff4f87367fcb8e2f339052c

i.e., layout optimization improves perf for both static and dynamic, plus dynamic is even a bit better (which seems to be the expectation)

<img width="557" height="159" alt="Image" src="https://github.com/user-attachments/assets/888b39ac-f309-48ea-a4c9-9ac96ae138ba" />
